projects/healthPingValidation.kp/knowledge.md

```diff
@@ -0,0 +1,343 @@
+
+
+
+
+    count       15.000000
+    mean     50329.800000
+    std      20164.844933
+    min       6475.000000
+    25%      37564.500000
+    50%      56195.000000
+    75%      59652.500000
+    95%      76980.000000
+    max      80578.000000
+    dtype: float64
+
+
+
+# Compute how many clients are reporting
+
+
+```python
+def getClients(pings):
+    clients = pings.map(lambda ping: ping["clientId"]).distinct()
+    return clients
+```
+
+```python
+clientsCount = getClients(cachedData).count()
+print "Clients number = " + str(clientsCount)
+```
+    Clients number = 67028
+
+
+# Compute average number of pings per client per day
+
+This include showing diagrams and printings stats
+
+
+
+```python
+from collections import Counter
+
+def getAvgPerDate(iterable):
+    aggregare = Counter(iterable)
+    result = sum(aggregare.values()) * 1.0 / len(aggregare)
+    return result
+    
+def pingsPerClientPerDay(pings):
+    idDateRDD = pings.map(lambda ping: (ping["clientId"], ping["meta/submissionDate"])).groupByKey()
+    return idDateRDD.map(lambda pair: (pair[0], getAvgPerDate(pair[1]))).collect()
+
+resultDistributionDict = pingsPerClientPerDay(cachedData)
+```
+
+```python
+values = map(lambda p: p[1], resultDistributionDict)
+
+plt.title("Average number of pings per client per day")
+plt.plot(sorted(values))
+
+pingsPerDaySeries = pd.Series(values)
+pingsPerDaySeries.describe([.25, .5, .75, .95])
+
+```
+
+
+
+
+    count    67028.000000
+    mean         3.041343
+    std         12.124944
+    min          1.000000
+    25%          1.333333
+    50%          2.000000
+    75%          3.222222
+    95%          7.090909
+    max       1278.000000
+    dtype: float64
+
+
+
+
+
+![png](images/output_21_1.png)
+
+
+# Although pings could be sent only 1 time in a hour (at max 24 times per day), some clients submitted more than 24 pings. 
+Half of this number could be described because of shutdown ping (TelemetrySend save this ping on shutdown and submit as pending in the next session, which could lead to loats of ping)
+
+# Daily active Health ping clients against DAU
+
+
+```python
+DAU = [('20170730', 28097), ('20170728', 33078), ('20170802', 36030), ('20170807', 40198), ('20170808', 37232),
+      ('20170806', 32071), ('20170801', 37161), ('20170729', 27523), ('20170727', 34326), ('20170809', 4376), 
+      ('20170810', 21), ('20170803', 36399), ('20170804', 36638), ('20170805', 32871), ('20170731', 34131)]
+```
+
+```python
+def getClientsPerDay(pings):
+    idDateRDD = pings.map(lambda ping: (ping["meta/submissionDate"], ping["clientId"])).groupByKey()
+    return idDateRDD.map(lambda pair: (pair[0], len(set(pair[1]))))
+
+clientsPerDay = getClientsPerDay(cachedData).collect()
+```
+
+```python
+plotlistofTuples(sorted(DAU), inColor='red')
+plotlistofTuples(sorted(clientsPerDay))
+plt.legend(['DAU', 'Health ping clients/day'])
+```
+
+
+
+
+    <matplotlib.legend.Legend at 0x7ff554e3b550>
+
+
+
+
+
+![png](images/output_26_1.png)
```

Comment:
The ratio of clients that send health pings is very high.
We need to dig more into that in a follow-up bug.
Specifically we need to
1) understand whats going on here
2) see if the beta shows the same behavior
3) see if we need to change the implementation
4) check if all this breaks our ping volume / storage projections (and if thats ok)